File: src/Hash.BaaS.Gateway.Sdk/GatewayRequestSigner.cs
using System.Globalization;
using System.Security.Cryptography;
using System.Text;
using NSec.Cryptography;

namespace Hash.BaaS.Gateway.Sdk;

/// <summary>
/// Signs outbound HTTP requests using the Hash BaaS Gateway RFC 9421 profile.
/// </summary>
public static class GatewayRequestSigner
{
    private const string Algorithm = "ed25519";
    private const string Tag = "hash-baas-v1";
    private const string SignatureLabel = "sig1";

    /// <summary>
    /// Mutates <paramref name="request"/> by adding required Gateway headers and signature headers.
    /// </summary>
    public static async Task SignAsync(
        HttpRequestMessage request,
        GatewaySdkOptions options,
        CancellationToken cancellationToken = default)
    {
        ArgumentNullException.ThrowIfNull(request);
        ArgumentNullException.ThrowIfNull(options);
        options.Validate();

        if (request.RequestUri is null || !request.RequestUri.IsAbsoluteUri)
            throw new InvalidOperationException("RequestUri must be an absolute URI before signing.");

        var isBodyBearing = IsUnsafeMethod(request);

        UpsertHeader(request, "X-Product-Code", options.ProductCode);
        UpsertHeader(request, "X-Client-Id", options.ClientId);
        UpsertHeader(request, "X-Audit-Source-Type", options.AuditSourceType);
        UpsertHeader(request, "X-Audit-User-Id", options.AuditUserId);

        string? contentDigest = null;
        string? contentType = null;
        string? idempotencyKey = null;

        if (isBodyBearing)
        {
            request.Content ??= new StringContent(string.Empty, Encoding.UTF8, "application/json");

            var bodyBytes = await request.Content!.ReadAsByteArrayAsync(cancellationToken).ConfigureAwait(false);
            contentDigest = $"sha-256=:{Convert.ToBase64String(SHA256.HashData(bodyBytes))}:";
            UpsertHeader(request, "Content-Digest", contentDigest);

            contentType = request.Content.Headers.ContentType?.ToString() ?? string.Empty;
            request.Content = CloneContent(request.Content, bodyBytes);

            if (!TryGetHeader(request, "Idempotency-Key", out idempotencyKey))
            {
                idempotencyKey = Guid.NewGuid().ToString();
                request.Headers.TryAddWithoutValidation("Idempotency-Key", idempotencyKey);
            }
        }

        var componentNames = BuildComponentNames(isBodyBearing);
        var created = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        var expires = created + options.MaxSignatureLifetimeSeconds;

        var parameters = new Dictionary<string, string>(StringComparer.Ordinal)
        {
            ["created"] = created.ToString(CultureInfo.InvariantCulture),
            ["expires"] = expires.ToString(CultureInfo.InvariantCulture),
            ["keyid"] = options.GetKeyId(),
            ["nonce"] = GenerateNonce(),
            ["alg"] = Algorithm,
            ["tag"] = Tag
        };

        var signatureBase = BuildSignatureBase(
            request,
            componentNames,
            parameters,
            contentDigest,
            contentType,
            idempotencyKey,
            options);

        byte[] signatureBytes;
        using (var key = Key.Import(
            SignatureAlgorithm.Ed25519,
            Encoding.UTF8.GetBytes(options.PrivateKeyPem),
            KeyBlobFormat.PkixPrivateKeyText))
        {
            signatureBytes = SignatureAlgorithm.Ed25519.Sign(key, Encoding.UTF8.GetBytes(signatureBase));
        }

        request.Headers.Remove("Signature-Input");
        request.Headers.Remove("Signature");
        request.Headers.TryAddWithoutValidation("Signature-Input", $"{SignatureLabel}={SerializeSignatureParams(componentNames, parameters)}");
        request.Headers.TryAddWithoutValidation("Signature", $"{SignatureLabel}=:{Convert.ToBase64String(signatureBytes)}:");
    }

    private static ByteArrayContent CloneContent(HttpContent source, byte[] bodyBytes)
    {
        var clone = new ByteArrayContent(bodyBytes);
        foreach (var header in source.Headers)
            clone.Headers.TryAddWithoutValidation(header.Key, header.Value);

        return clone;
    }

    private static bool IsUnsafeMethod(HttpRequestMessage request)
    {
        var method = request.Method;
        return method == HttpMethod.Post
            || method == HttpMethod.Put
            || method == HttpMethod.Patch
            || (method == HttpMethod.Delete && request.Content is not null);
    }

    private static List<string> BuildComponentNames(bool isBodyBearing)
    {
        var names = new List<string> { "@method", "@target-uri", "@authority" };
        if (isBodyBearing)
        {
            names.Add("content-digest");
            names.Add("content-type");
        }

        names.Add("x-product-code");
        names.Add("x-client-id");
        names.Add("x-audit-source-type");
        names.Add("x-audit-user-id");

        if (isBodyBearing)
            names.Add("idempotency-key");

        return names;
    }

    private static void UpsertHeader(HttpRequestMessage request, string name, string value)
    {
        request.Headers.Remove(name);
        request.Headers.TryAddWithoutValidation(name, value);
    }

    private static bool TryGetHeader(HttpRequestMessage request, string name, out string value)
    {
        if (request.Headers.TryGetValues(name, out var values))
        {
            value = string.Join(", ", values).Trim();
            return !string.IsNullOrWhiteSpace(value);
        }

        value = string.Empty;
        return false;
    }

    private static string GenerateNonce()
    {
        var bytes = new byte[16];
        RandomNumberGenerator.Fill(bytes);
        return Convert.ToBase64String(bytes)
            .TrimEnd('=')
            .Replace('+', '-')
            .Replace('/', '_');
    }

    private static string BuildSignatureBase(
        HttpRequestMessage request,
        IReadOnlyList<string> componentNames,
        IReadOnlyDictionary<string, string> parameters,
        string? contentDigest,
        string? contentType,
        string? idempotencyKey,
        GatewaySdkOptions options)
    {
        var sb = new StringBuilder();
        foreach (var name in componentNames)
        {
            var value = ResolveComponentValue(request, name, contentDigest, contentType, idempotencyKey, options);
            sb.Append('"').Append(name).Append("\": ").Append(value).Append('\n');
        }

        sb.Append("\"@signature-params\": ");
        sb.Append(SerializeSignatureParams(componentNames, parameters));
        return sb.ToString();
    }

    private static string ResolveComponentValue(
        HttpRequestMessage request,
        string name,
        string? contentDigest,
        string? contentType,
        string? idempotencyKey,
        GatewaySdkOptions options)
    {
        return name switch
        {
            "@method" => request.Method.Method.ToUpperInvariant(),
            "@target-uri" => request.RequestUri?.AbsoluteUri ?? string.Empty,
            "@authority" => BuildAuthority(request.RequestUri),
            "content-digest" => contentDigest ?? string.Empty,
            "content-type" => contentType ?? string.Empty,
            "x-product-code" => options.ProductCode,
            "x-client-id" => options.ClientId,
            "x-audit-source-type" => options.AuditSourceType,
            "x-audit-user-id" => options.AuditUserId,
            "idempotency-key" => idempotencyKey ?? string.Empty,
            _ => TryGetHeader(request, name, out var headerValue) ? headerValue : string.Empty
        };
    }

    private static string BuildAuthority(Uri? uri)
    {
        if (uri is null)
            return string.Empty;

        var host = uri.Host.ToLowerInvariant();
        var defaultPort = uri.Scheme switch
        {
            "https" => 443,
            "http" => 80,
            _ => -1
        };

        return uri.Port == defaultPort || uri.IsDefaultPort
            ? host
            : $"{host}:{uri.Port}";
    }

    private static string SerializeSignatureParams(
        IReadOnlyList<string> componentNames,
        IReadOnlyDictionary<string, string> parameters)
    {
        var sb = new StringBuilder();
        sb.Append('(');
        for (var i = 0; i < componentNames.Count; i++)
        {
            if (i > 0)
                sb.Append(' ');

            sb.Append('"').Append(componentNames[i]).Append('"');
        }

        sb.Append(')');

        foreach (var kvp in parameters)
        {
            sb.Append(';').Append(kvp.Key).Append('=');
            if (kvp.Key is "created" or "expires")
                sb.Append(kvp.Value);
            else
                sb.Append('"').Append(kvp.Value).Append('"');
        }

        return sb.ToString();
    }
}

File: java/src/main/java/ge/hashbank/baas/gateway/GatewayRequestSigner.java
package ge.hashbank.baas.gateway;

import java.net.URI;
import java.net.http.HttpRequest;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.security.Signature;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.UUID;

public final class GatewayRequestSigner {
    private static final String SIGNATURE_LABEL = "sig1";
    private static final String ALGORITHM = "ed25519";
    private static final String TAG = "hash-baas-v1";
    private static final SecureRandom RANDOM = new SecureRandom();

    private GatewayRequestSigner() {
    }

    public static HttpRequest sign(String method, URI uri, byte[] body, String contentType, GatewaySdkOptions options) {
        var unsafe = isUnsafe(method, body);
        var requestBody = unsafe ? (body == null ? new byte[0] : body) : null;

        var builder = HttpRequest.newBuilder(uri)
                .header("Accept", "application/json")
                .header("X-Product-Code", options.productCode())
                .header("X-Client-Id", options.clientId())
                .header("X-Audit-Source-Type", options.auditSourceType())
                .header("X-Audit-User-Id", options.auditUserId());

        String contentDigest = null;
        String idempotencyKey = null;
        String effectiveContentType = contentType == null || contentType.isBlank()
                ? "application/json"
                : contentType;

        if (unsafe) {
            contentDigest = "sha-256=:" + Base64.getEncoder().encodeToString(sha256(requestBody)) + ":";
            idempotencyKey = UUID.randomUUID().toString();
            builder.header("Content-Digest", contentDigest)
                    .header("Content-Type", effectiveContentType)
                    .header("Idempotency-Key", idempotencyKey);
        }

        var components = buildComponents(unsafe);
        var created = Instant.now().getEpochSecond();
        var parameters = new LinkedHashMap<String, String>();
        parameters.put("created", Long.toString(created));
        parameters.put("expires", Long.toString(created + options.maxSignatureLifetimeSeconds()));
        parameters.put("keyid", options.keyId());
        parameters.put("nonce", nonce());
        parameters.put("alg", ALGORITHM);
        parameters.put("tag", TAG);

        var signatureBase = buildSignatureBase(method, uri, components, parameters, contentDigest, effectiveContentType, idempotencyKey, options);
        var signature = sign(signatureBase, options.privateKeyPem());
        builder.header("Signature-Input", SIGNATURE_LABEL + "=" + serializeSignatureParams(components, parameters))
                .header("Signature", SIGNATURE_LABEL + "=:" + Base64.getEncoder().encodeToString(signature) + ":");

        if (unsafe) {
            builder.method(method.toUpperCase(Locale.ROOT), HttpRequest.BodyPublishers.ofByteArray(requestBody));
        } else {
            builder.method(method.toUpperCase(Locale.ROOT), HttpRequest.BodyPublishers.noBody());
        }

        return builder.build();
    }

    private static boolean isUnsafe(String method, byte[] body) {
        var upper = method.toUpperCase(Locale.ROOT);
        return upper.equals("POST")
                || upper.equals("PUT")
                || upper.equals("PATCH")
                || (upper.equals("DELETE") && body != null);
    }

    private static ArrayList<String> buildComponents(boolean unsafe) {
        var components = new ArrayList<String>();
        components.add("@method");
        components.add("@target-uri");
        components.add("@authority");
        if (unsafe) {
            components.add("content-digest");
            components.add("content-type");
        }
        components.add("x-product-code");
        components.add("x-client-id");
        components.add("x-audit-source-type");
        components.add("x-audit-user-id");
        if (unsafe) {
            components.add("idempotency-key");
        }
        return components;
    }

    private static String buildSignatureBase(
            String method,
            URI uri,
            ArrayList<String> components,
            LinkedHashMap<String, String> parameters,
            String contentDigest,
            String contentType,
            String idempotencyKey,
            GatewaySdkOptions options) {
        var base = new StringBuilder();
        for (var component : components) {
            base.append('"').append(component).append("\": ")
                    .append(resolve(component, method, uri, contentDigest, contentType, idempotencyKey, options))
                    .append('\n');
        }
        base.append("\"@signature-params\": ").append(serializeSignatureParams(components, parameters));
        return base.toString();
    }

    private static String resolve(
            String component,
            String method,
            URI uri,
            String contentDigest,
            String contentType,
            String idempotencyKey,
            GatewaySdkOptions options) {
        return switch (component) {
            case "@method" -> method.toUpperCase(Locale.ROOT);
            case "@target-uri" -> uri.toString();
            case "@authority" -> authority(uri);
            case "content-digest" -> contentDigest;
            case "content-type" -> contentType;
            case "x-product-code" -> options.productCode();
            case "x-client-id" -> options.clientId();
            case "x-audit-source-type" -> options.auditSourceType();
            case "x-audit-user-id" -> options.auditUserId();
            case "idempotency-key" -> idempotencyKey;
            default -> "";
        };
    }

    private static String authority(URI uri) {
        var host = uri.getHost().toLowerCase(Locale.ROOT);
        var port = uri.getPort();
        var scheme = uri.getScheme();
        if (port == -1 || (scheme.equals("https") && port == 443) || (scheme.equals("http") && port == 80)) {
            return host;
        }
        return host + ":" + port;
    }

    private static String serializeSignatureParams(ArrayList<String> components, LinkedHashMap<String, String> parameters) {
        var result = new StringBuilder("(");
        for (var i = 0; i < components.size(); i++) {
            if (i > 0) {
                result.append(' ');
            }
            result.append('"').append(components.get(i)).append('"');
        }
        result.append(')');
        for (var entry : parameters.entrySet()) {
            result.append(';').append(entry.getKey()).append('=');
            if (entry.getKey().equals("created") || entry.getKey().equals("expires")) {
                result.append(entry.getValue());
            } else {
                result.append('"').append(entry.getValue()).append('"');
            }
        }
        return result.toString();
    }

    private static byte[] sign(String signatureBase, String privateKeyPem) {
        try {
            var signer = Signature.getInstance("Ed25519");
            signer.initSign(GatewayKeyGenerator.importPrivateKey(privateKeyPem));
            signer.update(signatureBase.getBytes(StandardCharsets.UTF_8));
            return signer.sign();
        } catch (Exception ex) {
            throw new IllegalStateException("Failed to sign Gateway request", ex);
        }
    }

    private static byte[] sha256(byte[] value) {
        try {
            return MessageDigest.getInstance("SHA-256").digest(value);
        } catch (Exception ex) {
            throw new IllegalStateException("SHA-256 is not available", ex);
        }
    }

    private static String nonce() {
        var bytes = new byte[16];
        RANDOM.nextBytes(bytes);
        return Base64.getUrlEncoder().withoutPadding().encodeToString(bytes);
    }
}
